python/pyspark/sql/types.py

```diff
@@ -305,7 +305,7 @@ def jsonValue(self):
     @classmethod
     def fromJson(cls, json):
```

Comment:
Just to clarify, does this cause any user-facing error? As long as we can roundtrip jsonValue and fromJson, I think we're good.

Reply:
Nope it won't cause any user-facing error. By doing:
```python
json.get("containsNull", True)
```
The code is getting the value of the `containsNull` key from the JSON if it exists. If the key does not, it will be `True`. This matches the constructor of `ArrayType` which has `True` as the default argument for `containsNull`.

I'm happy to add some tests if you believe it is needed but this is be a no-op change.

Reply:
@HyukjinKwon I refactored the test to include a roundtrip test for `jsonValue()` 👍

Reply:
In which case `containsNull` does not exist?

Reply:
The first test case:
```python
(
  "ArrayType with default values", ArrayType,
  {"elementType": "string"},
  ArrayType(StringType())
 ),
```
Shows that is `containsNull` is not populated in the incoming JSON it will behave the same as if the argument is not given to the `ArrayType` constructor.

Reply:
It has:

```python
>>> ArrayType(StringType()).jsonValue()
{'type': 'array', 'elementType': 'string', 'containsNull': True}
```

Reply:
Right so the default value for `containsNull` for `ArrayType` is `True` so this test shows that without supplying it in the JSON or Constructor you get the same result. I will add another `assert` for the resulting JSON.

Reply:
I mean I am trying to understand why you want this change. For example, does it affect anything in the roundtrip between `jsonValue` and `fromJson`, or are you trying to build up the JSON by yourself somewhere?

Reply:
Added!

Reply:
Correct we have a use case where we build up the JSON elsewhere and we don't want to have to require the default keys.

Reply:
I don't think the format is meant to be exposed from end users. It's just for internal ser/de purpose that is supposed to be roundtrip between `jsonValue` and `fromJson` to workaround Py4J limitation.

BTW, here isn't only place where you should handle JSON ser/de. For example, you should also change Scala side at `DataType.parseDataType`.

Reply:
@HyukjinKwon I think going down the rabbit hole of use case is a bad approach. Taking a step back from the use case, the motivation behind this change is for a **consistent API when creating DataTypes** (namely `StructField`, `ArrayType`, & `MapType`). It seems a bit inconsistent that I can do:
```python
>> MapType(StringType(), StringType()) // notice I omit the valueContainsNull argument
MapType(StringType,StringType,true) // valueContainsNull has a default value of True
```
But when using `.fromJson` which is essentially a constructor for `MapType`, with an equivalent JSON
```python
>> MapType.fromJson({"keyType": "string", "valueType": "string"}) // No valueContainsNull argument
KeyError: 'valueContainsNull' // Error thrown tell me I need it in my JSON
```
The same inconsistencies hold true for `StructField` keys `metadata` & `nullable` **_AND_** `ArrayType` key `containsNull`.

On the flip side, lets say I omit a required parameter:
```python
>> MapType(StringType()) // notice I omit the valueType parameter
TypeError: __init__() missing 1 required positional argument: 'valueType' // Error throw telling me I need valueType
```
And when I do the same using a JSON constructing:
```
>> MapType({"keyType": "string"}) // omit the valueType parameter
KeyError: 'valueType' // Error throw telling me I need it in my JSON
``

This change proposes making the constructors, both the Python class constructor and the classMethod `.fromJson`, consistent on what they expect for input.

Reply:
The JSON format itself isn't supposed to be manually constructed. I think it doesn't necessarily have to be consistent with the constructor.

Reply:
> The JSON format itself isn't supposed to be manually constructed

I see where you're coming from here but the function `.fromJson` for for `ArrayType`, `MapType`, `StructField`, and `StructType` is exposed to developers so it _can_ be used. I have a strong hunch people do use it as it is an easy way to generate and store schemas. I'm curious what it would take to get this through? It's _seems_ like a fairly logical change.

Reply:
Unless there are plans to remove `.fromJson`, it is a publicly exposed interface and, I dare say, a rather useful one.

JSON is currently the only schema definition structure that is a) human readable, b) machine readable without `exec`, and c) easy to generate with anything other than Python / Java.

As far as I can tell, this PR:
- Adds additional test coverage for an existing component
- Makes an existing component more flexible for some use cases
- Does not reduce any existing functionality

Getting to use cases, I have frequently found value in providing a machine readable schema that can be validated with JSON schema and used as a unit and integration tests to verify expected schema against a SQL file.